lightning/src/offers/flow.rs

```diff
@@ -0,0 +1,1658 @@
+	/// [`Event::PaymentFailed`]: crate::events::Event::PaymentFailed
+	/// [Avoiding Duplicate Payments]: #avoiding-duplicate-payments
+	pub fn pay_for_offer(
+		&self, offer: &Offer, quantity: Option<u64>, amount_msats: Option<u64>,
+		payer_note: Option<String>, payment_id: PaymentId, retry_strategy: Retry,
+		max_total_routing_fee_msat: Option<u64>,
+	) -> Result<(), Bolt12SemanticError> {
+		self.pay_for_offer_intern(
+			offer,
+			quantity,
+			amount_msats,
+			payer_note,
+			payment_id,
+			None,
+			|invoice_request, nonce| {
+				let expiration = StaleExpiration::TimerTicks(1);
+				let retryable_invoice_request =
+					RetryableInvoiceRequest { invoice_request: invoice_request.clone(), nonce };
+				self.commons
+					.add_new_awaiting_invoice(
+						payment_id,
+						expiration,
+						retry_strategy,
+						max_total_routing_fee_msat,
+						Some(retryable_invoice_request),
+					)
+					.map_err(|_| Bolt12SemanticError::DuplicatePaymentId)
+			},
+		)
+	}
+
+	/// Creates a [`Bolt12Invoice`] for a [`Refund`] and enqueues it to be sent via an onion
+	/// message.
+	///
+	/// The resulting invoice uses a [`PaymentHash`] recognized by the [`ChannelManager`] and a
+	/// [`BlindedPaymentPath`] containing the [`PaymentSecret`] needed to reconstruct the
+	/// corresponding [`PaymentPreimage`]. It is returned purely for informational purposes.
+	///
+	/// [`PaymentPreimage`]: crate::types::payment::PaymentPreimage
+	///
+	/// # Limitations
+	///
+	/// Requires a direct connection to an introduction node in [`Refund::paths`] or to
+	/// [`Refund::payer_signing_pubkey`], if empty. This request is best effort; an invoice will be
+	/// sent to each node meeting the aforementioned criteria, but there's no guarantee that they
+	/// will be received and no retries will be made.
+	///
+	/// # Errors
+	///
+	/// Errors if:
+	/// - the refund is for an unsupported chain, or
+	/// - the parameterized [`Router`] is unable to create a blinded payment path or reply path for
+	///   the invoice.
+	///
+	/// [`BlindedPaymentPath`]: crate::blinded_path::payment::BlindedPaymentPath
+	/// [`ChannelManager`]: crate::ln::channelmanager::ChannelManager
+	/// [`PaymentHash`]: crate::types::payment::PaymentHash
+	/// [`PaymentSecret`]: crate::types::payment::PaymentSecret
+	/// [`Router`]: crate::routing::router::Router
+	pub fn request_refund_payment(
+		&self, refund: &Refund,
+	) -> Result<Bolt12Invoice, Bolt12SemanticError> {
+		let expanded_key = &self.inbound_payment_key;
+		let entropy = &*self.entropy_source;
+		let secp_ctx = &self.secp_ctx;
+
+		let amount_msats = refund.amount_msats();
+		let relative_expiry = DEFAULT_RELATIVE_EXPIRY.as_secs() as u32;
+
+		if refund.chain() != self.commons.get_chain_hash() {
+			return Err(Bolt12SemanticError::UnsupportedChain);
+		}
+
+		// TODO: Add persistance through `commons` internal function. Shouldn't be exposed here.
+		// let _persistence_guard = PersistenceNotifierGuard::notify_on_drop(&*self.commons);
```

Comment:
@TheBlueMatt

Hi Matt! I had disabled this here because I couldn’t cleanly separate its usage from the OffersMessageCommons trait. I’m curious about why we needed persistence here—could we move it further down in the code, potentially pairing it with the function that originally required this guard?

Reply:
Removed in [pr3412.07](https://github.com/shaavan/rust-lightning/commits/pr3412.07)
Thanks for the pointer! ✨